### PR TITLE
Prefs [Task][26302] Create TimeTracking View Presenter Contract

### PR DESCRIPTION
**Pull request by toandv98**
## Related Tickets

- [#Task 26302](https://edu-redmine.sun-asterisk.vn/issues/26302)

## WHAT

- Create TimeTracking View Presenter Contract

## HOW

- Create View Presenter Contract, TimeTrackingFragment, TimeTracingPresenter implement Contract

## Review Checklist

| Category      | View Point                                                   | Description                                                  | Expected Reviewer Answer | Reviewer1 (ToanDV) | Reviewer2 (HuyTQ)   |
| ------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------ | ------------------ | ------------------ |
| General       | Code should be self-explanatory.                             | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Is this function or class too big? If yes, is the function or class having too many responsibilities? |                                                              | NO                       | <li>- [x] no</li>  | <li>- [x] no</li>  |
| General       | Does the code follow DRY?: Do not Repeat Yourself            | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Exception handling and cleanup (dispose) resources           |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Does the code build and execute correctly?                   |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Conventions   | Does the source code have end-line errors?                   |                                                              | NO                       | <li>- [x] no</li>  | <li>- [x] no</li>  |
| Conventions   | No hard coding, use constants/configuration values           |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Conventions   | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Redmine       | Does the ticket follow Sun* Redmine working process?         | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |

## Notes (Optional)

*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/res/layout/fragment_time_tracking.xml [hunk 1]
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <include
+        android:id="@+id/include"
```
Comment (thonc-2135) [inline app/src/main/res/layout/fragment_time_tracking.xml]: tên ý nghĩa hơn chút đi

File: app/src/main/res/layout/item_time_record.xml [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    <androidx.constraintlayout.widget.Guideline
+        android:id="@+id/guideline2"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:orientation="horizontal"
+        app:layout_constraintGuide_begin="@dimen/dp_80" />
+
+    <TextView
+        android:id="@+id/text_label_timein"
```
Comment (thonc-2135) [inline app/src/main/res/layout/item_time_record.xml]: sai convention, trên thì đặt tên camel case dưới thì lại kiểu underscore r

File: app/src/main/java/com/edu/sun/oereminder/utils/DateTimeExtension.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+fun GregorianCalendar.toTime(): String {
+    return if (timeInMillis == 0L) TIME_NULL else SimpleDateFormat(
+        TIME_FORMAT,
+        Locale.getDefault()
+    ).format(time)
+}
+
+fun GregorianCalendar.toDate(pattern: String): String {
+    return SimpleDateFormat(pattern, Locale.getDefault()).format(time)
+}
```
Comment (thonc-2135) [inline app/src/main/java/com/edu/sun/oereminder/utils/DateTimeExtension.kt]:
```suggestion
fun GregorianCalendar.toDate(pattern: String) = SimpleDateFormat(pat...)
```

File: app/src/main/java/com/edu/sun/oereminder/ui/base/BaseFragment.kt [hunk 4]
```diff
@@ -31,6 +31,11 @@ abstract class BaseFragment<V : BaseView, T : BasePresenterImpl<V>> : Fragment()
         initListener()
     }
 
+    override fun onDestroyView() {
```
Comment (thonc-2135) [inline app/src/main/java/com/edu/sun/oereminder/ui/base/BaseFragment.kt:38]: chú ý thứ tự hàm , nên theo thứ tứ của vòng đời fragment

File: app/src/main/java/com/edu/sun/oereminder/ui/report/ReportContract.kt [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package com.edu.sun.oereminder.ui.report
+
+import com.edu.sun.oereminder.ui.base.BasePresenter
+import com.edu.sun.oereminder.ui.base.BaseView
+
+interface ReportContract {
+    interface View : BaseView {
+
+    }
```
Comment (thonc-2135) [inline app/src/main/java/com/edu/sun/oereminder/ui/report/ReportContract.kt]:
```suggestion
    interface View : BaseView {
    }
```
t thấy nhiều file ô cách r nhiều file lại cách thêm 1 dòng như v. bỏ đi cho file n ngắn bớt

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeRecordAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                    textPartOfDay.text = partOfDay
+                }
+            }
+        }
+    }
+
+    class TimeRecordItemCallBack : DiffUtil.ItemCallback<TimeRecord>() {
+        override fun areItemsTheSame(oldItem: TimeRecord, newItem: TimeRecord): Boolean {
+            return oldItem.id == newItem.id
+        }
```
Comment (thonc-2135) [inline app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeRecordAdapter.kt]:
```suggestion
        override fun areItemsTheSame(oldItem: TimeRecord, newItem: TimeRecord) =
            oldItem.id == newItem.id

```

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeRecordAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                    textStatus.text = if (status.isBlank()) {
+                        context.getString(R.string.status_pending)
+                    } else status
+                    textPartOfDay.text = partOfDay
+                }
+            }
+        }
+    }
+
+    class TimeRecordItemCallBack : DiffUtil.ItemCallback<TimeRecord>() {
```
Comment (tuanpa-0908) [inline app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeRecordAdapter.kt]: Cho ra file riêng đi ô

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingFragment.kt [hunk 8]
```diff
@@ -18,9 +29,62 @@ class TimeTrackingFragment : BaseFragment<View, TimeTrackingPresenter>(), View {
...
+
+        picker = MaterialDatePicker.Builder.dateRangePicker().apply {
+            setCalendarConstraints(
+                CalendarConstraints.Builder().setEnd(GregorianCalendar().timeInMillis).build()
+            )
+        }.build()
+
+        recyclerTimeTracking.apply {
+            adapter = timeTrackingAdapter
+            layoutManager = LinearLayoutManager(context)
```
Comment (tuanpa-0908) [inline app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingFragment.kt]: ô có thể set layout manager ngay trong xml cho gọn nhẹ

File: app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingFragment.kt [hunk 9]
```diff
@@ -18,9 +29,62 @@ class TimeTrackingFragment : BaseFragment<View, TimeTrackingPresenter>(), View {
...
+        }
     }
 
     override fun initListener() {
 
+        btnDatePicker.setOnClickListener {
+            activity?.let {
+                picker?.show(it.supportFragmentManager, picker.toString())
+            }
```
Comment (tuanpa-0908) [inline app/src/main/java/com/edu/sun/oereminder/ui/timetracking/TimeTrackingFragment.kt]: VIết được trên 1 dòng không nhỉ

File: app/src/main/java/com/edu/sun/oereminder/utils/DateTimeExtension.kt [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+const val DATE_FORMAT = "dd MMMM"
+const val DATE_FORMAT_DAY_OF_WEEK = "EEEE, dd-MM-yyyy"
+const val TIME_FORMAT = "hh:mm"
+const val TIME_NULL = "--:--"
+
+fun GregorianCalendar.toDate(): String {
+    return SimpleDateFormat(DATE_FORMAT_DAY_OF_WEEK, Locale.getDefault()).format(time)
+}
+
+fun GregorianCalendar.toTime(): String {
```
Comment (tuanpa-0908) [inline app/src/main/java/com/edu/sun/oereminder/utils/DateTimeExtension.kt]:
```suggestion
fun GregorianCalendar.toTime() = if ...
```

**Pull request merged**